File: Python/src/sensor_reader.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
温湿度传感器读取模块
封装Modbus通讯和传感器数据读取逻辑
支持多个串口连接
"""

from typing import Tuple, Optional, Dict, Any, List
from pymodbus.client import ModbusSerialClient
from pymodbus.exceptions import ModbusException


class ModbusSensorReader:
    """Modbus温湿度传感器读取器"""
    
    def __init__(
        self,
        port: str,
        baudrate: int = 9600,
        parity: str = 'N',
        stopbits: int = 1,
        bytesize: int = 8,
        timeout: float = 1.0
    ):
        """
        初始化传感器读取器
        
        :param port: 串口号
        :param baudrate: 波特率
        :param parity: 校验位（N/E/O）
        :param stopbits: 停止位
        :param bytesize: 数据位
        :param timeout: 超时时间（秒）
        """
        self.port = port
        self.baudrate = baudrate
        self.parity = parity
        self.stopbits = stopbits
        self.bytesize = bytesize
        self.timeout = timeout
        
        self.client: Optional[ModbusSerialClient] = None
    
    def connect(self) -> bool:
        """建立与Modbus设备的连接"""
        self.client = ModbusSerialClient(
            port=self.port,
            baudrate=self.baudrate,
            parity=self.parity,
            stopbits=self.stopbits,
            bytesize=self.bytesize,
            timeout=self.timeout
        )
        return self.client.connect()
    
    def disconnect(self):
        """关闭连接"""
        if self.client:
            self.client.close()
            self.client = None
    
    def read_sensor_data(
        self,
        slave_id: int,
        temp_register: int,
        humidity_register: int,
        temp_scale: float = 0.1,
        humidity_scale: float = 0.1
    ) -> Optional[Tuple[float, float]]:
        """
        读取单个传感器的温湿度数据
        
        :param slave_id: 从站地址
        :param temp_register: 温度寄存器地址
        :param humidity_register: 湿度寄存器地址
        :param temp_scale: 温度缩放因子
        :param humidity_scale: 湿度缩放因子
        :return: (温度, 湿度) 元组，失败返回None
        """
        if not self.client:
            print("错误：未建立连接")
            return None
        
        try:
            temp_result = self.client.read_holding_registers(
                temp_register, 1, slave_id=slave_id
            )
            if temp_result.isError():
                print(f"传感器 {slave_id} 读取温度失败: {temp_result}")
                return None
            
            humi_result = self.client.read_holding_registers(
                humidity_register, 1, slave_id=slave_id
            )
            if humi_result.isError():
                print(f"传感器 {slave_id} 读取湿度失败: {humi_result}")
                return None
            
            temp_raw = temp_result.registers[0]
            if temp_raw > 32767:
                temperature = (temp_raw - 65536) / temp_scale
            else:
                temperature = temp_raw / temp_scale
            
            humidity_raw = humi_result.registers[0]
            humidity = humidity_raw / humidity_scale
            
            return temperature, humidity
            
        except ModbusException as e:
            print(f"Modbus异常 (传感器 {slave_id}): {e}")
            return None
        except Exception as e:
            print(f"读取传感器 {slave_id} 数据失败: {e}")
            return None


class MultiPortReader:
    """多串口Modbus传感器读取器"""
    
    def __init__(self, ports_config: List[Dict[str, Any]]):
        """
        初始化多串口读取器
        
        :param ports_config: 串口配置列表，每个配置包含port, baudrate等参数
        """
        self.readers: Dict[str, ModbusSensorReader] = {}
        self.port_configs: Dict[str, Dict[str, Any]] = {}
        
        for config in ports_config:
            port_name = config.get("name", config["port"])
            port = config["port"]
            self.port_configs[port_name] = config
            self.readers[port_name] = ModbusSensorReader(
                port=port,
                baudrate=config.get("baudrate", 9600),
                parity=config.get("parity", "N"),
                stopbits=config.get("stopbits", 1),
                bytesize=config.get("bytesize", 8),
                timeout=config.get("timeout", 1.0)
            )
    
    def connect_all(self) -> bool:
        """连接所有串口"""
        success_count = 0
        for name, reader in self.readers.items():
            if reader.connect():
                print(f"已连接到串口: {name} ({reader.port})")
                success_count += 1
            else:
                print(f"连接串口失败: {name} ({reader.port})")
        return success_count > 0
    
    def disconnect_all(self):
        """关闭所有连接"""
        for reader in self.readers.values():
            reader.disconnect()
        print("已关闭所有串口连接")
    
    def get_reader(self, port_name: str) -> Optional[ModbusSensorReader]:
        """获取指定串口的读取器"""
        return self.readers.get(port_name)
    
    def read_sensor_data(
        self,
        port_name: str,
        slave_id: int,
        temp_register: int,
        humidity_register: int,
        temp_scale: float = 0.1,
        humidity_scale: float = 0.1
    ) -> Optional[Tuple[float, float]]:
        """
        读取指定串口传感器的数据
        
        :param port_name: 串口名称
        :param slave_id: 从站地址
        :param temp_register: 温度寄存器地址
        :param humidity_register: 湿度寄存器地址
        :param temp_scale: 温度缩放因子
        :param humidity_scale: 湿度缩放因子
        :return: (温度, 湿度) 元组，失败返回None
        """
        reader = self.readers.get(port_name)
        if not reader:
            print(f"错误：未找到串口 {port_name}")
            return None
        return reader.read_sensor_data(
            slave_id, temp_register, humidity_register, temp_scale, humidity_scale
        )


def read_all_sensors(
    reader: MultiPortReader,
    sensors: list
) -> Dict[str, Optional[Tuple[float, float]]]:
    """
    批量读取所有传感器数据
    
    :param reader: MultiPortReader实例
    :param sensors: 传感器配置列表，每个传感器需包含port字段
    :return: 传感器名称到数据的映射
    """
    results = {}
    for sensor in sensors:
        name = sensor["name"]
        port_name = sensor.get("port", "default")
        slave_id = sensor["slave_id"]
        temp_reg = sensor.get("temp_reg", 0)
        humi_reg = sensor.get("humi_reg", 1)
        temp_scale = sensor.get("temp_scale", 0.1)
        humi_scale = sensor.get("humi_scale", 0.1)
        
        data = reader.read_sensor_data(
            port_name, slave_id, temp_reg, humi_reg, temp_scale, humi_scale
        )
        results[name] = data
    
    return results

File: C/src/sensor_reader.c
#include "sensor_reader.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#ifdef _WIN32
    #include <windows.h>
    #include "modbus.h"
#else
    #include <unistd.h>
    #include <modbus.h>
#endif

struct ModbusSensorReader {
    char name[64];
    char port[64];
    int baudrate;
    int data_bits;
    int stop_bits;
    char parity;
    float timeout;
    modbus_t* ctx;
};

struct MultiPortReader {
    int port_count;
    ModbusSensorReader* ports[16];
};

ModbusSensorReader* modbus_reader_create(const char* port, int baudrate, int data_bits, int stop_bits, char parity, float timeout) {
    ModbusSensorReader* reader = (ModbusSensorReader*)malloc(sizeof(ModbusSensorReader));
    if (!reader) return NULL;

    strncpy(reader->port, port, sizeof(reader->port) - 1);
    reader->port[sizeof(reader->port) - 1] = '\0';
    reader->baudrate = baudrate;
    reader->data_bits = data_bits;
    reader->stop_bits = stop_bits;
    reader->parity = parity;
    reader->timeout = timeout;
    reader->ctx = NULL;
    strncpy(reader->name, "默认串口", sizeof(reader->name) - 1);
    reader->name[sizeof(reader->name) - 1] = '\0';

    return reader;
}

bool modbus_reader_connect(ModbusSensorReader* reader) {
    if (!reader) return false;

    reader->ctx = modbus_new_rtu(
        reader->port,
        reader->baudrate,
        reader->parity,
        reader->data_bits,
        reader->stop_bits
    );

    if (!reader->ctx) {
        fprintf(stderr, "无法创建Modbus RTU上下文: %s\n", modbus_strerror(errno));
        return false;
    }

    modbus_set_debug(reader->ctx, FALSE);
    modbus_set_response_timeout(reader->ctx, (uint32_t)(reader->timeout * 1000000), 0);

    if (modbus_connect(reader->ctx) == -1) {
        fprintf(stderr, "连接失败: %s\n", modbus_strerror(errno));
        modbus_free(reader->ctx);
        reader->ctx = NULL;
        return false;
    }

    return true;
}

void modbus_reader_free(ModbusSensorReader* reader) {
    if (reader) {
        if (reader->ctx) {
            modbus_close(reader->ctx);
            modbus_free(reader->ctx);
        }
        free(reader);
    }
}

bool modbus_reader_read_sensor(ModbusSensorReader* reader, uint8_t slave_id, uint16_t temp_reg, uint16_t humi_reg, float temp_scale, float humi_scale, SensorData* result) {
    if (!reader || !reader->ctx || !result) return false;

    result->error = false;
    result->temperature = 0.0f;
    result->humidity = 0.0f;
    result->error_msg[0] = '\0';

    modbus_set_slave(reader->ctx, slave_id);

    uint16_t temp_data[1];
    int temp_rc = modbus_read_input_registers(reader->ctx, temp_reg, 1, temp_data);

    if (temp_rc == -1) {
        result->error = true;
        snprintf(result->error_msg, sizeof(result->error_msg), "读取温度失败: %s", modbus_strerror(errno));
        return false;
    }

    uint16_t humi_data[1];
    int humi_rc = modbus_read_input_registers(reader->ctx, humi_reg, 1, humi_data);

    if (humi_rc == -1) {
        result->error = true;
        snprintf(result->error_msg, sizeof(result->error_msg), "读取湿度失败: %s", modbus_strerror(errno));
        return false;
    }

    result->temperature = (float)temp_data[0] * temp_scale;
    result->humidity = (float)humi_data[0] * humi_scale;

    return true;
}

MultiPortReader* multi_port_reader_create(void) {
    MultiPortReader* mreader = (MultiPortReader*)malloc(sizeof(MultiPortReader));
    if (!mreader) return NULL;

    mreader->port_count = 0;
    for (int i = 0; i < 16; i++) {
        mreader->ports[i] = NULL;
    }

    return mreader;
}

bool multi_port_reader_add_port(MultiPortReader* mreader, const char* name, const char* port, int baudrate, int data_bits, int stop_bits, char parity, float timeout) {
    if (!mreader || mreader->port_count >= 16) return false;

    ModbusSensorReader* reader = modbus_reader_create(port, baudrate, data_bits, stop_bits, parity, timeout);
    if (!reader) return false;

    if (name) {
        strncpy(reader->name, name, sizeof(reader->name) - 1);
        reader->name[sizeof(reader->name) - 1] = '\0';
    }

    mreader->ports[mreader->port_count] = reader;
    mreader->port_count++;

    return true;
}

bool multi_port_reader_connect_all(MultiPortReader* mreader) {
    if (!mreader) return false;

    int success_count = 0;
    for (int i = 0; i < mreader->port_count; i++) {
        if (mreader->ports[i]) {
            if (modbus_reader_connect(mreader->ports[i])) {
                printf("成功连接到串口 %s: %s\n", mreader->ports[i]->name, mreader->ports[i]->port);
                success_count++;
            } else {
                printf("连接到串口 %s 失败: %s\n", mreader->ports[i]->name, mreader->ports[i]->port);
            }
        }
    }

    return success_count > 0;
}

void multi_port_reader_free(MultiPortReader* mreader) {
    if (mreader) {
        for (int i = 0; i < mreader->port_count; i++) {
            if (mreader->ports[i]) {
                modbus_reader_free(mreader->ports[i]);
            }
        }
        free(mreader);
    }
}

ModbusSensorReader* multi_port_reader_find_port(MultiPortReader* mreader, const char* port_name) {
    if (!mreader || !port_name) return NULL;

    for (int i = 0; i < mreader->port_count; i++) {
        if (mreader->ports[i] && strcmp(mreader->ports[i]->name, port_name) == 0) {
            return mreader->ports[i];
        }
    }

    return NULL;
}

void multi_port_reader_read_sensors(MultiPortReader* mreader, const SensorConfig* sensors, int sensor_count, SensorData* results) {
    if (!mreader || !sensors || !results) return;

    for (int i = 0; i < sensor_count; i++) {
        results[i].error = false;
        results[i].temperature = 0.0f;
        results[i].humidity = 0.0f;
        results[i].error_msg[0] = '\0';

        strncpy(results[i].name, sensors[i].name, sizeof(results[i].name) - 1);
        results[i].name[sizeof(results[i].name) - 1] = '\0';

        ModbusSensorReader* reader = multi_port_reader_find_port(mreader, sensors[i].port_name);
        if (!reader) {
            results[i].error = true;
            snprintf(results[i].error_msg, sizeof(results[i].error_msg), "未找到串口: %s", sensors[i].port_name);
            continue;
        }

        if (!reader->ctx) {
            results[i].error = true;
            snprintf(results[i].error_msg, sizeof(results[i].error_msg), "串口未连接: %s", sensors[i].port_name);
            continue;
        }

        if (!modbus_reader_read_sensor(reader, sensors[i].slave_id, sensors[i].temp_register, sensors[i].humi_register, sensors[i].temp_scale, sensors[i].humi_scale, &results[i])) {
            results[i].error = true;
        }
    }
}

void print_sensor_data(const SensorData* results, int count) {
    if (!results) return;

    for (int i = 0; i < count; i++) {
        if (results[i].error) {
            printf("  %s: 读取失败 - %s\n", results[i].name, results[i].error_msg);
        } else {
            printf("  %s: 温度=%.1f°C, 湿度=%.1f%%\n", results[i].name, results[i].temperature, results[i].humidity);
        }
    }
}
